### PR TITLE
Thread-safe operator[], configurable number of mutexes

### LINKED ISSUE #18: Improve locking for vectors/matrices?
Currently, we are locking the whole matrix for both read and write operations. This may limits parallelism too much. 
There are two things we could investigate too improve on the situation:
1. Use ```shared_mutex``` ([C++17](http://en.cppreference.com/w/cpp/thread/shared_mutex) or [Boost](http://www.boost.org/doc/libs/1_63_0/doc/html/thread/synchronization.html#thread.synchronization.mutex_types.shared_mutex)) to allow for parallel reads
2. Do not lock the whole matrix for operations like ```set_entry``` that only affect a single entry. This could be done by using several mutexes (maybe one mutex per row?)

### PR DESCRIPTION
**Pull request by tobiasleibner**
A first proposal to fix #18 and #20. I only adapted the Eigen vectors for now, but the same should work for the other vectors and matrices. 
I am following the lines of http://www.gotw.ca/gotw/045.htm, so I removed the locks in all methods that yield undefined behavior when called from two threads at the same time anyway (e.g. when calling ``vector.set_entry(0, 0.)`` from one thread and ``vector.set_entry(0, 1.)`` from another thread at the same time, we can't tell if the result is 0 or 1 even if ``set_entry`` is thread-safe).
To make it possible to call ``set_entry`` and ``add_to_entry`` for different entries in parallel, the vectors now take a ``num_mutexes`` argument (defaulted to 1) that determines the number of mutexes used. If ``num_mutexes == 10`` and ``vector.size() == 100``, entries 0-9 will be protected by the first mutex, entries 10-19 by the second and so on. This greatly reduced locking for me when using the L2 projection in dune-gdt, which calls ``set_entry`` on the vector of the discrete function repeatedly on each entitity.
Having several mutexes implies that we need to lock all of them when doing global operations like ``axpy`` which might worsen performance when doing a lot of global operations, so I kept the number of mutexes configurable.
As for ``operator[]``, I added an ``unshareable_`` flag that is set when the non-const ``operator[]`` is called. If that flag is set, we will always do a deep copy instead of copy-on-write for that vector.

### CODE REVIEW

File: dune/xt/la/container/common.hh [hunk 1]
```diff
@@ -287,9 +303,13 @@ protected:
...
   inline ScalarType& operator[](const size_t ii)
   {
+    ensure_uniqueness();
+    unshareable_ = true;
     return get_entry_ref(ii);
   }
 
+  // TODO: Need to ensure uniqueness() and unshareable_ = true also in this method, but that would imply making
+  // ensure_uniqueness const and all members mutable
   inline const ScalarType& operator[](const size_t ii) const
```
Comment (ftalbrecht) [inline dune/xt/la/container/common.hh]: We could also think about changing the interface of the const `operator[]` to always return a copy, though I have no feeling for the impact...
```c++
   inline const ScalarType operator[](const size_t ii) const
```

File: dune/xt/la/container/common.hh [hunk 2]
```diff
@@ -720,6 +789,7 @@ public:
     }
   }
 
+  // TODO: Why is this constructor needed? Remove?
```
Comment (ftalbrecht) [inline dune/xt/la/container/common.hh]: Remove! (left over from old python bindings)

File: dune/xt/la/container/common.hh [hunk 3]
```diff
@@ -762,49 +837,63 @@ public:
   } // CommonSparseMatrix(...)
 
   template <int ROWS, int COLS>
-  explicit operator Dune::FieldMatrix<ScalarType, ROWS, COLS>() const
+  explicit operator std::unique_ptr<Dune::FieldMatrix<ScalarType, ROWS, COLS>>() const
```
Comment (ftalbrecht) [inline dune/xt/la/container/common.hh]: Could you comment on why you return a `std::unique_ptr` here instead of the value (thus having a different signature than the `operator` below)? I woul think that with RTO we could just return by value and create the unique_ptr on the outside, if required...

File: dune/xt/la/container/common.hh [hunk 4]
```diff
@@ -762,49 +837,63 @@ public:
...
   }
 
   /// \}
   /// \name Required by ContainerInterface.
   /// \{
-
   inline ThisType copy() const
   {
     ThisType ret(*this);
+    ret.entries_ = std::make_shared<EntriesVectorType>(*entries_);
```
Comment (ftalbrecht) [inline dune/xt/la/container/common.hh]: Why is this required? Is this not exactly what happens in the copy ctor above?

File: dune/xt/la/container/common.hh (hunk 3 above)
Comment (tobiasleibner) [inline dune/xt/la/container/common.hh]: For larger matrices (something like 1000*1000) creating a FieldMatrix on the stack causes a stack-overflow. This happened to me and just returning a unique_ptr seemed to be the fastest solution. I am not sure how else to avoid the stack-overflow. Maybe we can just add back the old operator without the unique_ptr and also keep the new one, so we can use the old operator whenever possible and explicitly use the unique_ptr one if the matrix is too large.

File: dune/xt/la/container/common.hh (hunk 4 above)
Comment (tobiasleibner) [inline dune/xt/la/container/common.hh]: In the copy constructor, a shallow copy is done unless ``unshareable_ == true``, and the ``copy()`` method has to do a deep copy. If `` unshareable_ == true`` we will copy two times, so maybe I can add a check for that.

**Comment by tobiasleibner** [comment]
Concerning the ``get_entry_ref`` methods, I now removed the guards from the vectors. I am not sure about the matrices though, as we do not have an ``operator[]`` there, do we want to expose the ``get_entry_ref`` methods as a replacement (and thus use guards?). On the one hand it might be confusing if ``get_entry_ref`` is not safe for vectors but safe for matrices. But on the other hand it would also be nice to have a method to get a reference to an entry safely also for matrices.

File: dune/xt/la/container/common.hh (hunk 2 above)
Comment (tobiasleibner) [inline dune/xt/la/container/common.hh]: Done

File: dune/xt/la/container/common.hh (hunk 3 above)
Comment (ftalbrecht) [inline dune/xt/la/container/common.hh]: Ok, makes sense! Nevertheless, I would add the operator without the ptr, to be usable for smaller matrices...

**Comment by tobiasleibner** [comment]
I now renamed the ``protected`` method ``get_entry_ref`` to ``get_unchecked_ref``. The ``get_entry_ref`` method is now ``public`` both for vectors and matrices and is equivalent to ``operator[]`` for vectors, i.e. once it is called copy-on-write will be disabled for the vector/matrix.
This is ready to merge now if the tests are passing and you do not have objections.

**Comment by tobiasleibner** [comment]
Note to myself: Add tests for new ``operator[]`` behavior before merging, check iterators.
